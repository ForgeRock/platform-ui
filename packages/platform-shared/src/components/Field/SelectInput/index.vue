<!-- Copyright (c) 2020-2024 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <FrInputLayout
    :id="internalId"
    :name="name"
    :description="description"
    :errors="combinedErrors"
    :is-html="isHtml"
    :label="label"
    :floating-label="floatingLabel"
    :readonly-label="disabled">
    <FrMultiselectBase
      ref="vms"
      :id="internalId"
      :model-value="inputValue"
      @update:modelValue="inputHandler"
      v-bind="$attrs"
      :option-height="optionHeightCalculation"
      :class="classes"
      :data-testid="testid"
      :testid="testid"
      :name="name"
      :options="selectOptions"
      track-by="value"
      :disabled="disabled"
      :placeholder="internalPlaceholder"
      :searchable="searchable"
      :internal-search="internalSearch"
      :show-labels="false"
      :allow-empty="allowEmpty"
      :autofocus="autofocus"
      label="text"
      :combobox-labelledby="inputLabelledby"
      @select="$emit('select', $event)"
      @open="openHandler"
      @close="closeHandler"
      @tag="$emit('tag', $event)"
      @search-change="$emit('search-change', $event)">
      <template #noResult>
        {{ $t('common.noResult') }}
      </template>
      <template
        v-for="(key, slotName) in $slots"
        #[slotName]="slotData">
        <!-- @slot passthrough slot -->
        <slot
          :name="slotName"
          v-bind="slotData" />
      </template>
    </FrMultiselectBase>
    <template
      v-for="(key, slotName) in $slots"
      #[slotName]="slotData">
      <!-- @slot passthrough slot -->
      <slot
        :name="slotName"
        v-bind="slotData" />
    </template>
  </FrInputLayout>
</template>

<script>
import {
  find, isEqual,
} from 'lodash';
import {
  computed, watch, ref, toRef,
} from 'vue';
import { useField } from 'vee-validate';
import uuid from 'uuid/v4';
import FrMultiselectBase from '../../MultiselectBase/MultiselectBase';
import i18n from '@/i18n';
import FrInputLayout from '../Wrapper/InputLayout';
import { useGetId, useValueIsDifferent } from '../Wrapper/InputComposable';

export default {
  name: 'SelectInput',
  components: {
    FrInputLayout,
    FrMultiselectBase,
  },
  props: {
    allowEmpty: {
      type: Boolean,
      default: false,
    },
    /**
     * Autofocus field when rendered.
     */
    autofocus: {
      type: Boolean,
      default: false,
    },

    /**
     * Related text that displays underneath field.
     */
    description: {
      type: String,
      default: '',
    },
    /**
     * if field should be disabled.
     */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * List of errors related to input value (mostly used for callback components)
     */
    errors: {
      type: Array,
      default: () => [],
    },
    /**
     * Boolean to show a floating label or above label on controls
     */
    floatingLabel: {
      type: Boolean,
      default: true,
    },
    /**
     * Unique ID
     */
    id: {
      type: String,
      default: '',
    },
    /**
     * For accessibility: if you want to manually associate the input field with a label, legend etc.
     */
    inputLabelledby: {
      type: String,
      default: '',
    },
    /**
     * Search through existing options
     */
    internalSearch: {
      type: Boolean,
      default: true,
    },
    /**
     * Boolean to render label and help text as html.
     */
    isHtml: {
      type: Boolean,
      default: false,
    },
    /**
     * Label value that is show as placeholder value on floating labels or static label above control in other case.
     */
    label: {
      type: String,
      default: '',
    },
    /**
     * Input name.
     */
    name: {
      type: String,
      required: true,
      default: () => uuid(),
    },
    /**
     * Height of the individual option items. Important to
     * set accurately so the options menu aligns correctly.
     */
    optionHeightCalculation: {
      type: Number,
      default: 40,
    },
    /**
     * Options for select input.
     */
    options: {
      type: [Array, Object],
      required: true,
    },
    placeholder: {
      type: String,
      default: '',
    },
    /**
     * Boolean to show the input as readonly.
     */
    readonly: {
      type: Boolean,
      default: false,
    },
    searchable: {
      type: Boolean,
      default: true,
    },
    /**
     * Optionally scrolls the selected option into view when the select is opened.
     */
    showSelectedOptionOnOpen: {
      type: Boolean,
      default: false,
    },
    /**
     * Optionally sorts the displayed select options by their text attribute.
     */
    sortOptions: {
      type: Boolean,
      default: false,
    },
    testid: {
      type: String,
      default: null,
    },
    /**
     * Vee-validate validation types to check against.
     */
    validation: {
      type: [String, Object],
      default: '',
    },
    /**
     * Whether error validation should happen when this component renders.
     */
    validationImmediate: {
      type: Boolean,
      default: false,
    },
    /**
     * v-model value
     */
    value: {
      type: [Array, Object, Number, String, Boolean],
      default: '',
    },

  },
  setup(props, context) {
    const vms = ref(null);
    const internalId = useGetId(props);
    const isOpen = ref(false);

    const selectOptions = computed(() => {
      let formattedOptions;

      if (props.options.length && Object.hasOwnProperty.call(props.options[0], 'value')) {
        formattedOptions = [...props.options];
      } else {
        formattedOptions = props.options.map((option) => {
          const formattedOption = typeof (option) === 'string' ? option.trim() : option;
          return {
            text: formattedOption,
            value: formattedOption,
          };
        });
      }

      if (props.sortOptions) {
        formattedOptions.sort((a, b) => a.text.localeCompare(b.text));
      }
      return formattedOptions;
    });

    function setFloatLabels(isOpenArg, value) {
      if (props.floatingLabel) {
        if (props.searchable && isOpenArg && !!props.label) {
          return true;
        }
        return (!!value || value === 0) && !!props.label;
      }
      return false;
    }

    /**
     * Process value prop to match with select options
     * Checks for the option in the select options and to see if new value is actually new
     *
     * @param {Object} newVal new input value
     * @returns {Object} input value
     */
    function processValue(newVal, old) {
      if (newVal !== undefined && newVal !== null) {
        const optionBeingSet = find(selectOptions.value, { value: newVal });
        if (optionBeingSet || useValueIsDifferent(old, newVal)) {
          return optionBeingSet;
        }
      }
      return undefined;
    }

    const {
      value: inputValue, errors: fieldErrors,
    } = useField(() => `${props.name}-id-${uuid()}`, toRef(props, 'validation'), {
      validateOnMount: props.validationImmediate,
      initialValue: processValue(props.value, undefined),
      bails: false,
    });

    const hasPrependBtn = ref(Object.keys(context.slots).includes('prependButton'));
    const floatLabels = ref(setFloatLabels(isOpen.value, inputValue.value));

    const internalPlaceholder = computed(() => (props.placeholder || props.searchable ? i18n.global.t('common.typeToSearch') : ''));
    const combinedErrors = computed(() => [...props.errors, ...fieldErrors.value]);
    const classes = computed(() => [
      { 'polyfill-placeholder': floatLabels.value },
      { 'h-100': props.floatingLabel },
      { 'has-prepend-button': hasPrependBtn.value },
      { 'no-multiselect-label': !props.label },
      'form-control', 'p-0', 'text-nowrap',
    ]);

    function openHandler() {
      context.emit('open');
      isOpen.value = true;
      floatLabels.value = setFloatLabels(true, inputValue.value);
    }

    function closeHandler(value) {
      context.emit('closed');
      isOpen.value = false;
      floatLabels.value = setFloatLabels(false, value);
    }

    function inputHandler(e) {
      inputValue.value = e;
    }

    watch(() => props.options, (newOptions, oldOptions) => {
      if (props.value) {
        const oldValueObject = oldOptions.find(({ value }) => value === props.value);
        const newValueObject = newOptions.find(({ value }) => value === props.value);
        if (!oldValueObject || (newValueObject && (oldValueObject.value === newValueObject.value && oldValueObject.text !== newValueObject.text))) {
          inputValue.value = processValue(props.value, inputValue.value);
        }
      }
    }, { deep: true });

    watch(() => inputValue.value, (value) => {
      const newValue = value !== null
        && value !== undefined
        && typeof value === 'object'
        && Object.hasOwnProperty.call(value, 'value')
        ? value.value : value;

      floatLabels.value = setFloatLabels(isOpen.value, newValue);
      context.emit('input', newValue);
    });

    watch(() => props.value, (value) => {
      const newValues = processValue(value);
      if (!isEqual(inputValue.value, newValues)) {
        inputValue.value = newValues;
      }
    });

    return {
      classes,
      closeHandler,
      combinedErrors,
      fieldErrors,
      floatLabels,
      inputHandler,
      inputValue,
      internalId,
      internalPlaceholder,
      openHandler,
      selectOptions,
      vms,
    };
  },
};
</script>

<style lang="scss" scoped>
.polyfill-placeholder {
  :deep(.multiselect__tags) {
    padding: 1.1rem 50px 0.1rem 0.75rem;
  }
  :deep(.multiselect__single) {
    margin-top: 4px;
    min-height: 26px;
  }
}
</style>
