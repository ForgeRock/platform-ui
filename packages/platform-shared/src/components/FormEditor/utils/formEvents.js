/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { cloneDeep } from 'lodash';

/**
 * Executes a given script in a Web Worker, providing form values and schema as variables.
 * The worker exists only to run the script and return the potentially modified form values and schema.
 * Worker is terminated after recieving the result.
 *
 * @param {string} script - The JavaScript code to execute inside the Web Worker.
 * @param {Object} formValues - The current form values to be passed to the worker.
 * @param {Object} formSchema - The form schema to be passed to the worker.
 * @returns {Promise<{formValues: Object, formSchema: Object}>} Resolves with the potentially updated form values and schema.
 * @throws {Error} If the worker encounters an error during script execution.
 */
// eslint-disable-next-line import/prefer-default-export
export async function useWebWorker(script, formValues, formSchema) {
  return new Promise((resolve, reject) => {
    const worker = new Worker(new URL('./formEventsWorker.js', import.meta.url));

    worker.onmessage = (e) => {
      if (e.data.error) {
        reject(new Error(`Error executing onLoad event: ${e.data.error}`));
        worker.terminate();
        return;
      }

      resolve({
        formValues: e.data.formValues,
        formSchema: e.data.formSchema,
      });

      worker.terminate();
    };

    const scriptVariables = {
      formValues: cloneDeep(formValues),
      formSchema: cloneDeep(formSchema),
    };

    worker.postMessage({
      script, scriptVariables,
    });
  });
}
