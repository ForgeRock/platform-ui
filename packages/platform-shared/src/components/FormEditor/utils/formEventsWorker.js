/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { get, set } from 'lodash';

let _tempFormValues = {};
let _tempFormSchema = [];

export function _setTempFormValues(values) {
  _tempFormValues = values;
}
export function _setTempFormSchema(schema) {
  _tempFormSchema = schema;
}

/**
 * Retrieves a field object from the form schema by its model key.
 *
 * @param {string} key - The model key to search for in the form schema.
 * @returns {Object|null} The field object if found, otherwise null.
 */
export function _getFieldByKey(key) {
  return _tempFormSchema.flat().find((f) => f.model === key) || null;
}

/**
 * Functions passed to the script to manipulate form fields and values.
 */

/**
 * Disables a form field by its key.
 *
 * @param {string} key - The unique identifier of the field to disable.
 */
function disableField(key) {
  const field = _getFieldByKey(key);
  if (field) field.disabled = true;
}

/**
 * Enables a form field by its key.
 *
 * @param {string} key - The unique identifier for the form field to enable.
 */
function enableField(key) {
  const field = _getFieldByKey(key);
  if (field) field.disabled = false;
}

/**
 * Retrieves the label of a form field by its key.
 *
 * @param {string} key - The unique identifier for the form field.
 * @returns {(string|null)} The label of the field if found, otherwise null.
 */
function getLabel(key) {
  const field = _getFieldByKey(key);
  if (field) return field.label;
  return null;
}

/**
 * Retrieves the value associated with the specified key from the _tempFormValues object.
 *
 * @param {string} key - The key whose value should be retrieved.
 * @returns {*} The value associated with the given key, or undefined if the key does not exist.
 */
function getValue(key) {
  return get(_tempFormValues, key, undefined);
}

/**
 * Hides a form field by its key.
 *
 * @param {string} key - The unique identifier of the field to hide.
 */
function hideField(key) {
  const field = _getFieldByKey(key);
  if (field) field.showAlways = false;
}

/**
 * Show a form field by its key.
 *
 * @param {string} key - The unique identifier for the form field to show.
 */
function showField(key) {
  const field = _getFieldByKey(key);
  if (field) field.showAlways = true;
}

/**
 * Sets the label of a form field identified by the given key.
 *
 * @param {string} key - The unique identifier of the form field.
 * @param {string} label - The new label to assign to the form field.
 */
function setLabel(key, label) {
  const field = _getFieldByKey(key);
  if (field) field.label = label;
}

/**
 * Sets the value for a given key in the _tempFormValues object.
 *
 * @param {string} key - The key for which the value should be set.
 * @param {*} value - The value to assign to the specified key.
 */
function setValue(key, value) {
  set(_tempFormValues, key, value);
}

export const form = {
  disableField,
  enableField,
  getLabel,
  getValue,
  hideField,
  showField,
  setLabel,
  setValue,
  currentFieldValue: null,
  urlParams: null,
};

/**
 * Executes a given script in a Web Worker, providing form values and schema as variables.
 * Attempt to execute the script and return the potentially modified form values and schema.
 * Return an error if the script execution fails for any reason.
 */

// eslint-disable-next-line import/prefer-default-export
export function onmessage(e) {
  try {
    const { script, scriptVariables } = e.data;
    _setTempFormSchema(scriptVariables.formSchema);
    _setTempFormValues(scriptVariables.formValues);
    form.currentFieldValue = scriptVariables.newFieldValue;
    form.urlParams = new URLSearchParams(scriptVariables.windowSearch);

    // eslint-disable-next-line no-new-func
    const func = new Function('form', script);
    func(form);

    postMessage({ formValues: _tempFormValues, formSchema: _tempFormSchema });
  } catch (error) {
    postMessage({ error });
  }
}

global.onmessage = onmessage;
