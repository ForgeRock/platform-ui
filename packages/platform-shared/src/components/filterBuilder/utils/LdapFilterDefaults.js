/**
 * Copyright (c) 2022 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

export const ldapOperatorOptions = {
  All: { label: 'All', delimeter: 'and' },
  Any: { label: 'Any', delimeter: 'or' },
};

export const ldapDefaultConditionOptions = {
  Equals: { label: 'is', value: '=' },
  NotEquals: { label: 'is not', value: '!=' },
  Gte: { label: 'GTE (>=)', value: '>=' },
  Gt: { label: 'GT (>)', value: '!<=' },
  Lte: { label: 'LTE (<=)', value: '<=' },
  Lt: { label: 'LT (<)', value: '!>=' },
  BitwiseAnd: { label: 'Bitwise AND', value: ':1.2.840.113556.1.4.803:=' },
  BitwiseNand: { label: 'Bitwise NAND', value: '!:1.2.840.113556.1.4.803:=' },
  BitwiseOr: { label: 'Bitwise OR', value: ':1.2.840.113556.1.4.804:=' },
  BitwiseNor: { label: 'Bitwise NOR', value: '!:1.2.840.113556.1.4.804:=' },
};

export const conditionMap = {
  equal: '=',
  ge: '>=',
  le: '<=',
};

export function getTypeFromValue(value, properties) {
  if (!value) return 'string';
  const property = properties.find((prop) => prop.value === value);
  return property ? property.type : 'string';
}
