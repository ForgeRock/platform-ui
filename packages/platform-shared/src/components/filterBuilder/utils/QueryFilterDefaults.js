/**
 * Copyright (c) 2020-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import i18n from '@/i18n';

export const operatorOptions = {
  All: { label: i18n.global.t('filters.all'), delimeter: 'and' },
  Any: { label: i18n.global.t('filters.any'), delimeter: 'or' },
};

export const defaultConditionOptions = {
  Contains: { label: i18n.global.t('filters.contains'), value: 'co', type: ['string'] },
  DoesNotContain: { label: i18n.global.t('filters.doesNotContain'), value: '!co', type: ['string'] },
  Is: { label: i18n.global.t('filters.is'), value: 'eq', type: ['string', 'number', 'boolean'] },
  IsNot: { label: i18n.global.t('filters.isNot'), value: '!eq', type: ['string', 'number'] },
  IsPresent: { label: i18n.global.t('filters.isPresent'), value: 'pr', type: ['string', 'number', 'boolean'] },
  IsNotPresent: { label: i18n.global.t('filters.isNotPresent'), value: '!pr', type: ['string', 'number', 'boolean'] },
  StartsWith: { label: i18n.global.t('filters.startsWith'), value: 'sw', type: ['string'] },
  DoesNotStartsWith: { label: i18n.global.t('filters.doesNotStartWith'), value: '!sw', type: ['string'] },
  Gte: { label: i18n.global.t('filters.greaterThanEqual'), value: 'ge', type: ['number'] },
  Gt: { label: i18n.global.t('filters.greaterThan'), value: 'gt', type: ['number'] },
  Lte: { label: i18n.global.t('filters.lessThanEqual'), value: 'le', type: ['number'] },
  Lt: { label: i18n.global.t('filters.lessThan'), value: 'lt', type: ['number'] },
};

export const governanceConditionOptions = {
  Contains: { label: i18n.global.t('filters.contains'), value: 'contains', type: ['string', 'array'] },
  DoesNotContain: { label: i18n.global.t('filters.doesNotContain'), value: 'not_contains', type: ['string', 'array'] },
  Is: { label: i18n.global.t('filters.is'), value: 'equals', type: ['string', 'reference', 'boolean', 'number'] },
  IsNot: { label: i18n.global.t('filters.isNot'), value: 'not_equals', type: ['string', 'reference', 'number'] },
  StartsWith: { label: i18n.global.t('filters.startsWith'), value: 'starts_with', type: ['string'] },
  DoesNotStartsWith: { label: i18n.global.t('filters.endsWith'), value: 'ends_with', type: ['string'] },
  HasChanged: { label: i18n.global.t('filters.hasChanged'), value: 'has_changed', type: ['reference', 'string', 'number'] },
  HasNotChanged: { label: i18n.global.t('filters.hasNotChanged'), value: 'has_not_changed', type: ['reference', 'string', 'number'] },
  Gte: { label: i18n.global.t('filters.greaterThanEqual'), value: 'gte', type: ['number'] },
  Gt: { label: i18n.global.t('filters.greaterThan'), value: 'gt', type: ['number'] },
  Lte: { label: i18n.global.t('filters.lessThanEqual'), value: 'lte', type: ['number'] },
  Lt: { label: i18n.global.t('filters.lessThan'), value: 'lt', type: ['number'] },
};

export const temporalValueOptions = [
  { text: i18n.global.t('filters.previousValueOf'), value: 'before' },
  { text: i18n.global.t('filters.currentValueOf'), value: 'after' },
];

export function getTypeFromValue(value, properties) {
  if (!value) return 'string';
  const property = properties.find((prop) => prop.value === value);
  return property ? property.type : 'string';
}
