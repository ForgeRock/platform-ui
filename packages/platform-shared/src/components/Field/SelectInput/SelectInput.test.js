/**
 * Copyright (c) 2023-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { mount, flushPromises } from '@vue/test-utils';
import { findByTestId } from '@forgerock/platform-shared/src/utils/testHelpers';
import i18n from '@/i18n';
import SelectInput from './index';

describe('SelectInput', () => {
  const defaultProps = {
    name: 'stub-name',
    testid: 'stub-testid',
    options: [],
  };

  function setup(props, data) {
    return mount(SelectInput, {
      global: {
        plugins: [i18n],
      },
      props: {
        ...defaultProps,
        ...props,
      },
      data() {
        return data;
      },
    });
  }

  async function clickOnOptionNumber(number, wrapper) {
    const select = wrapper.find('.multiselect');
    const elements = () => select.findAll('.multiselect__option');

    select.trigger('click');
    elements()[number].trigger('click');
    await flushPromises();
  }

  it('SelectInput component process options prop from array', () => {
    const wrapper = setup({ options: ['a', 'b', 'c'] });

    const expected = [
      { text: 'a', value: 'a' },
      { text: 'b', value: 'b' },
      { text: 'c', value: 'c' },
    ];

    expect(wrapper.vm.selectOptions).toEqual(expected);
  });

  it('SelectInput component passes through options object prop', () => {
    const options = [
      { text: 'd', value: 'd' },
      { text: 'e', value: 'e' },
      { text: 'f', value: 'f' },
    ];

    const wrapper = setup({ options });

    expect(wrapper.vm.selectOptions).toEqual(options);
  });

  it('SelectInput component renders the options', () => {
    const wrapper = setup({ options: ['a', 'b', 'c'] });

    const select = wrapper.find('.multiselect');
    const elements = select.findAll('.multiselect__element');

    expect(select.exists()).toBe(true);
    expect(elements.length).toBe(3);
  });

  it('SelectInput component Closes Dropdown', () => {
    const wrapper = setup({ options: ['a', 'b', 'c'], label: 'testLabel' });
    const multiselectInput = findByTestId(wrapper, 'multi-select-input-stub-testid');

    multiselectInput.trigger('click');
    expect(wrapper.vm.floatLabels).toBe(true);
    multiselectInput.trigger('keydown.esc');
    expect(wrapper.vm.floatLabels).toBe(false);
  });

  it('SelectInput component Closes Dropdown without floating label', async () => {
    const wrapper = setup({ options: ['a', 'b', 'c'], label: 'testLabel', floatingLabel: false });

    expect(wrapper.vm.floatLabels).toBe(false);
    await wrapper.vm.$refs.vms.$emit('close');
    expect(wrapper.vm.floatLabels).toBe(false);
  });

  it('SelectInput component allows single selections', () => {
    const wrapper = setup({ options: ['a', 'b', 'c'] });

    const select = wrapper.find('.multiselect');
    const elements = () => select.findAll('.multiselect__option');

    select.trigger('click');
    elements()[1].trigger('click');
    expect(wrapper.vm.inputValue).toEqual({ text: 'b', value: 'b' });

    select.trigger('click');
    elements()[0].trigger('click');
    expect(wrapper.vm.inputValue).toEqual({ text: 'a', value: 'a' });
  });

  it('SelectInput passes through component slots', () => {
    const wrapper = mount(SelectInput, {
      global: {
        plugins: [i18n],
      },
      props: {
        ...defaultProps,
      },
      slots: {
        prepend: '<span class="test_prepend">prepend</span>',
        append: '<span class="test_append">append</span>', // Will match <slot name="FooBar" />,
      },
    });

    expect(wrapper.find('.test_prepend').exists()).toBe(true);
    expect(wrapper.find('.test_append').exists()).toBe(true);
  });

  it('will update the displayed selected value if the text for that option changes', async () => {
    const wrapper = setup({
      options: [
        { text: 'ayy', value: 'a' },
        { text: 'bee', value: 'b' },
        { text: 'cee', value: 'c' },
      ],
      closeOnSelect: true,
      value: 'b',
    });

    const select = wrapper.find('.multiselect');
    const elements = () => select.findAll('.multiselect__option');

    select.trigger('click');
    elements()[1].trigger('click');
    await flushPromises();

    // Check that the SelectInput shows the correct initial text
    expect(wrapper.find('.multiselect__single').element.value).toBe('bee');

    await wrapper.setProps({
      options: [
        { text: 'ayy', value: 'a' },
        { text: 'beegees?', value: 'b' },
        { text: 'cee', value: 'c' },
      ],
    });

    expect(wrapper.find('.multiselect__single').element.value).toBe('beegees?');
  });

  it('SelectInput removes float for blank values', async () => {
    const wrapper = setup({
      options: [
        { text: 'ayy', value: 'a' },
        { text: 'bee', value: 'b' },
      ],
      allowEmpty: true,
      label: 'test label',
    });

    await clickOnOptionNumber(0, wrapper);
    expect(wrapper.vm.floatLabels).toBe(true);

    await clickOnOptionNumber(0, wrapper);
    expect(wrapper.vm.floatLabels).toBe(false);
  });

  it('SelectInput removes float for blank values without floating label', () => {
    const wrapper = setup({
      options: [
        { text: 'ayy', value: 'a' },
        { text: 'bee', value: 'b' },
      ],
      floatingLabel: false,
    });

    wrapper.setProps({ value: 'sdf' });
    expect(wrapper.vm.floatLabels).toBe(false);
    wrapper.setProps({ value: '', options: [] });
    expect(wrapper.vm.floatLabels).toBe(false);
  });

  it('SelectInput is not autofocused on absence of prop "autofocus"', () => {
    const wrapper = mount(SelectInput, {
      global: {
        plugins: [i18n],
      },
      props: {
        ...defaultProps,
        autofocus: false,
      },
      attachTo: document.body,
    });

    const multiselect = findByTestId(wrapper, 'stub-testid');
    expect(multiselect.attributes('autofocus')).toBe(undefined);
  });

  it('SelectInput is autofocused on prop "autofocus"', async () => {
    const wrapper = mount(SelectInput, {
      global: {
        plugins: [i18n],
      },
      props: {
        ...defaultProps,
        autofocus: true,
      },
      attachTo: document.body,
    });
    await flushPromises();

    const multiselect = findByTestId(wrapper, 'stub-testid');
    expect(multiselect.attributes('autofocus')).toBe('true');
  });

  it('Maintains focus on the multiselect component after selecting an option', async () => {
    const wrapper = setup({ options: ['a', 'b', 'c'] });

    // The internals of vue-test-utils does not update the DOM with focused element states
    // when triggering event listeners, so it wasn't possible to test the actual DOM element
    // when it was focused, but we are able to deduce that the menu is still focused by being
    // able to immediately trigger it using a keyboard event after making a selection.
    const multiselect = findByTestId(wrapper, 'stub-testid');
    const multiselectInput = findByTestId(wrapper, 'multi-select-input-stub-testid');
    const options = wrapper.findAll('.multiselect__option');

    // Menu is opened initially
    multiselect.trigger('focus');

    // Option selected, menu closes, menu expected to remain focused
    options[0].trigger('click');
    expect(multiselectInput.attributes('aria-expanded')).toBe('false');

    // We can deduce that the menu is still focused because we can
    // immediately trigger it again using the down arrow.
    multiselect.trigger('keydown', { key: 'ArrowDown' });
    await wrapper.vm.$nextTick();
    expect(multiselectInput.attributes('aria-expanded')).toBe('true');
  });

  it('Displays the multiselect menu options if the "down arrow" is clicked', async () => {
    const wrapper = setup({ options: ['a', 'b', 'c'] });

    const multiselect = findByTestId(wrapper, 'stub-testid');
    const multiselectInput = findByTestId(wrapper, 'multi-select-input-stub-testid');

    await multiselect.trigger('keydown', { key: 'ArrowDown' });
    expect(multiselectInput.attributes('aria-expanded')).toBe('true');
  });

  it('Keeps the previously selected option highlighted after revealing the menu using the "up arrow" or "down arrow"', async () => {
    const wrapper = setup({ options: ['a', 'b', 'c'] });
    const multiselect = findByTestId(wrapper, 'stub-testid');
    const multiselectInput = findByTestId(wrapper, 'multi-select-input-stub-testid');
    const options = wrapper.findAll('.multiselect__option');

    // When the menu is initially opened, we expect the first option to always be highlighted
    await multiselect.trigger('focus');
    expect(options[0].classes('multiselect__option--highlight')).toBe(true);

    // The second option selected, menu closes, menu remains focused
    await options[1].trigger('click');
    expect(multiselectInput.attributes('aria-expanded')).toBe('false');

    // We trigger the menu to show again by clicking on the "down arrow" and the second option should now be highlighted
    await multiselect.trigger('keydown', { key: 'ArrowDown' });
    expect(multiselectInput.attributes('aria-expanded')).toBe('true');
    expect(options[0].classes('multiselect__option--highlight')).toBe(true);
    expect(options[1].classes('multiselect__option--selected')).toBe(true);

    // The third option selected, menu closes, menu remains focused
    await options[2].trigger('click');
    expect(multiselectInput.attributes('aria-expanded')).toBe('false');

    // We trigger the menu once more by clicking on the "up arrow" this time and the third option should remain highlighted
    await multiselect.trigger('keydown', { key: 'ArrowDown' });
    expect(multiselectInput.attributes('aria-expanded')).toBe('true');
    expect(options[0].classes('multiselect__option--highlight')).toBe(true);
    expect(options[1].classes('multiselect__option--highlight')).toBe(false);
    expect(options[2].classes('multiselect__option--selected')).toBe(true);
  });
});
