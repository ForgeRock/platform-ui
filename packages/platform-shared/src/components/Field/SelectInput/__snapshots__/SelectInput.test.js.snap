// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput @renders default: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation WATCH_ARRAY) "watch" option or vm.$watch on an array value will no longer trigger on array mutation unless the "deep" option is specified. If current usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ WATCH_ARRAY: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/watch.html",
    "
",
    " at <SelectInput",
    "name="stub-name"",
    "testid="stub-testid"",
    "options=",
    [],
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation WATCH_ARRAY) (2)",
    "
",
    " at <VueMultiselect",
    "id="floatingLabelInput1"",
    "ref="vms"",
    "value=""",
    " ...",
    ">",
    "
",
    " at <InputLayout",
    "id="floatingLabelInput1"",
    "name="stub-name"",
    "description=""",
    " ...",
    ">",
    "
",
    " at <SelectInput",
    "name="stub-name"",
    "testid="stub-testid"",
    "options=",
    [],
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "appear" with v-bind value \`false\` will render appear="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <Transition",
    "name="multiselect__loading"",
    ">",
    "
",
    " at <VueMultiselect",
    "id="floatingLabelInput1"",
    "ref="vms"",
    "value=""",
    " ...",
    ">",
    "
",
    " at <InputLayout",
    "id="floatingLabelInput1"",
    "name="stub-name"",
    "description=""",
    " ...",
    ">",
    "
",
    " at <SelectInput",
    "name="stub-name"",
    "testid="stub-testid"",
    "options=",
    [],
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "persisted" with v-bind value \`false\` will render persisted="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <Transition",
    "name="multiselect__loading"",
    ">",
    "
",
    " at <VueMultiselect",
    "id="floatingLabelInput1"",
    "ref="vms"",
    "value=""",
    " ...",
    ">",
    "
",
    " at <InputLayout",
    "id="floatingLabelInput1"",
    "name="stub-name"",
    "description=""",
    " ...",
    ">",
    "
",
    " at <SelectInput",
    "name="stub-name"",
    "testid="stub-testid"",
    "options=",
    [],
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_ENUMERATED_COERCION) Enumerated attribute "spellcheck" with v-bind value \`false\` will render the value as-is instead of coercing the value to "true" in Vue 3. Always use explicit "true" or "false" values for enumerated attributes. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_ENUMERATED_COERCION: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <VueMultiselect",
    "id="floatingLabelInput1"",
    "ref="vms"",
    "value=""",
    " ...",
    ">",
    "
",
    " at <InputLayout",
    "id="floatingLabelInput1"",
    "name="stub-name"",
    "description=""",
    " ...",
    ">",
    "
",
    " at <SelectInput",
    "name="stub-name"",
    "testid="stub-testid"",
    "options=",
    [],
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
